### PR TITLE
add the correctness checks for .NET Core 2.0/2.1/2.2 and full .NET framework

### LINKED ISSUE #257: Add CI correctness jobs for .NET, Core 2.0/2.1/2.2
In #246 I have introduced a bug which did not allow to run the benchmarks easily for .NET Core 2.0/2.1/2.2

We need to have a correctness check for not only 3.0 but for the other frameworks as well.

### PR DESCRIPTION
**Pull request by adamsitnik**
Fixes #257

### CODE REVIEW

**Comment by adamsitnik** [comment]
@adiaaida @jorive PTAL

obraz

File: scripts/dotnet.py [hunk 1]
```diff
@@ -181,10 +181,22 @@ def get_host_commit_sha(dotnet_path: str = None) -> str:
...
+        # Product Information:
+        #   Version:            2.1.202
+        #   Commit SHA-1 hash:  281caedada
         if 'Host' in decoded_line:
             foundHost = True
+        elif 'Product' in decoded_line:
+            foundHost = True
         elif foundHost and 'Commit' in decoded_line:
-            return decoded_line.strip().split()[1]
+            return decoded_line.strip().split()[-1]
```
Comment (adamsitnik) [inline scripts/dotnet.py]: [-1] is the last element. the string gets trimmed first, so this will always be our sha

File: azure-pipelines.yml [hunk 2]
```diff
@@ -65,11 +65,61 @@ jobs:
...
+              _Category: CoreCLR
+              _Framework: net461
+              _BuildConfig: $(_Architecture)_$(_Category)_$(_Framework)
+              _BenchViewArguments: ''
+            x86_CoreCLR_full_framework:
+              _Architecture: x86
+              _Category: CoreCLR
+              _Framework: net461
+              _BuildConfig: $(_Architecture)_$(_Category)_$(_Framework)
+              _BenchViewArguments: ''
```
Comment (adamsitnik) [inline azure-pipelines.yml]: @adiaaida I know that adding so many elements to the matrix looks a little bit lame, but I have no better idea. I am open to suggestions

Comment (adiaaida) [reply azure-pipelines.yml]: This is expected. I think it looks fine.

Only comment here is I think we should add some comments for like-blocks of definitions. So like above the netcoreapp3.0 ones, a comment saying # netcoreapp3.0 definitions; above the 2.2 # netcoreapp2.2, etc.

**Comment by adiaaida** [comment]
The only other thing I could think of is that there is an each keyword. And example can be found here: https://dnceng.visualstudio.com/internal/_git/dotnet-optimization?path=%2Feng%2Fcollect-optimization-data.yml&version=GBmaster&line=118&lineStyle=plain&lineEnd=119&lineStartColumn=1&lineEndColumn=1

Basically, that goes over the list of artifacts that we've specified in azure-pipelines.yml and pulls them down. The artifact list is specified here: https://dnceng.visualstudio.com/internal/_git/dotnet-optimization?path=%2Fazure-pipelines.yml&version=GBmaster&line=51&lineStyle=plain&lineEnd=52&lineStartColumn=1&lineEndColumn=1

However, I'm not sure if that would work for a matrix, which might not like each. But if we could have some parameter that had a list, say like this:

```
frameworks:
  - name: netcoreapp3.0
  - name: netcoreapp2.2
  - name: netcoreapp2.1
  - name: net461
```

and then above the matrix:

```
${{ each framework in frameworks }}:
  ,,,
  _Framework: ${{ framework.name }}
```

The problem is the naming of each matrix option, which I don't know if can be parameterized. @chcosta would have a better answer for that.

File: scripts/dotnet.py (hunk 1 above)
Comment (jorive) [reply scripts/dotnet.py]: The difference between using element in position 1 vs using the last element is that the current dotnet cli have two tokens and we select the second one as being the SHA. In the future, if this changes, we do not know what is the last token being selected. I would suggest leaving it as-is.

Comment (adamsitnik) [reply scripts/dotnet.py]: I think that using the last element is actually more future-proof

For .NET Core 2.0 we had:

```log
Commit SHA-1 hash:  281caedada
```

For .NET Core 2.1+ we have:

```log
Commit:  7a7ca06512
```

I think that if it's ever going to change, somebody is going to add text before the SHA, not after it. Also, now we have a CI that validates that, so if it ever breaks we will know about this with the first broken PR. Maestro sends few PRs a day so we are going to find this out very quickly ;)

**Comment by adamsitnik** [comment]
@adiaaida thank you for the suggestions! I tried to google for some more complex matrix examples but all I could find was one for debug and one for release or one for x64 and one for x86.

I am going to merge this as-is and add a comments and try to use the for loops in my next PR where I will add CoreFX to the list.

File: scripts/dotnet.py (hunk 1 above)
Comment (jorive) [reply scripts/dotnet.py]: Maybe moving forward we should use a regex to match this information. Something along the lines of: `^\W+Commit:\W+\b[0-9a-f]{5,40}\b$`.

Comment (adamsitnik) [reply scripts/dotnet.py]: that would be most probably the best solution!

**Pull request merged**